### PR TITLE
Smart 404!

### LINKED ISSUE #75: Smart 404 page?
I thought of this idea and implemented it on my own site, basically whenever the user encounters a 404 page, instead of just giving a 404 message it provides or even redirects to the likely correct url.  First it compares the distance between ```window.location.pathname``` and each of the correct paths on the website and returns the closest one. This uses the edit distance between them, using this snippet: 
```javascript
function distance(s1, s2) {
		s1 = s1.toLowerCase(); // Case doesn't matter in url's
		s2 = s2.toLowerCase();

		var costs = new Array();
		for (var i = 0; i <= s1.length; i++) {
			var lastValue = i;
			for (var j = 0; j <= s2.length; j++) {
				if (i == 0) costs[j] = j;
				else {
					if (j > 0) {
						var newValue = costs[j - 1];
						if (s1.charAt(i - 1) != s2.charAt(j - 1))
							newValue = Math.min(Math.min(newValue, lastValue), costs[j]) + 1;
						costs[j - 1] = lastValue;
						lastValue = newValue;
					}
				}
			}
			if (i > 0) costs[s2.length] = lastValue;
		}
		return costs[s2.length];
	}
```
Then, to get the closest url I do this: 
```js
function closest(str) {
		var urls = ['index', 'blog', 'projects']; //Correct url's
		urls.sort((a, b) => editDistance(a, str) - editDistance(str, b));
		return urls[0];
	}
```
Now I can check _how_ close these are, if they're 1 or 2 letters different than it's likely that the user meant that url:
```js
// url = window.location.pathname
if(editDistance(closest(url), url) < 3) {
    window.location.href = `http://example.com/${closest(url)}`;
  }
```
Either way this makes an amazing user experience on any website and we should definitely do this on Scratch Addons.

(Also you can try it [here](http://explosion.cf/indx) or for no redirect [here](http://explosion.cf/porjctes) )

### PR DESCRIPTION
**Pull request by Explosion-Scratch**

### CODE REVIEW

File: site/404.html [hunk 1]
```diff
@@ -8,10 +8,59 @@ <h1>404</h1>
...
+	document.getElementById(
+		'closest'
+	).innerHTML = `Did you mean "<a href='${newurl}.html'>${newurl}.html</a>"?`;
+	if (url === '404') {
+		document.getElementById('closest').innerHTML =
+			'Why would you purposely navigate to this page.';
+	}
+	// Functions
+	function closest(str) {
+		var urls = ['index', 'blog', 'projects'];
```
Comment (WorldLanguages) [inline site/404.html:39]: lol I think you're supposed to change this

**Review by WorldLanguages — APPROVED** [review]
I mean, it's fine but unnecessary. If someone rejects then I'm fine closing this PR.

**Comment by GrahamSH-LLK** [comment]
You should parse the sitemap to get the links

**Comment by Explosion-Scratch** [comment]
> I mean, it's fine but unnecessary. If someone rejects then I'm fine closing this PR.

But I _like_ it, I think it streamlines user experience quite a bit.

File: site/404.html (hunk 1 above)
Comment (Explosion-Scratch) [reply site/404.html:39]: Yeah, obviously XD. This is the array of correct pages.

Comment (Explosion-Scratch) [reply site/404.html:39]: > You should parse the sitemap to get the links

Wait, where do I find that?

**Comment by Hans5958** [comment]
Okay, here's my take. I'm not sure if this will scale well when, in the future, we will have lots of pages (even if using sitemap, which will give an unnecessary request, which will be large in the future), but I'll give it a pass this time.

The only thing that you need to do is to use the sitemap (https://scratchaddons.com/sitemap.xml) instead of using a stating array of pages.

**Comment by GrahamSH-LLK** [comment]
I can also add a sitemap.json, if that's easier to use.

**Comment by Hans5958** [comment]
To be honest, parsing `.xml` is already easy so another file is not needed.

**Review by williamhorning — APPROVED** [review]

**Comment by GrahamSH-LLK** [comment]
Please retarget (or redo in this case) these changes to the new website.

**Pull request closed without merging**